### ISSUE TITLE
Saving to and loading from external files

### ISSUE THREAD

**Issue by aorozsa** [haltu/muuri#314]
Here is my current saving/loading mechanism that's limited to within the session which works (where `savedItems` is a global variable):

`saveBtn.addEventListener('click', function(event) {`
`  savedItems = grid.getItems();`
`});`

`loadBtn.addEventListener('click', function(event) {`
`  grid.remove(grid.getItems(), {removeElements: true});`
`  savedItems.forEach(function(item) {`
`    grid.add(item.getElement());`
`  });`
`  ghost = grid.getItems(0)[0];`
`});`

I've already tried a couple of things to save the item element to a file, all of which have failed. First thing I've tried doing was stringifying `grid.getItems()`. There's something called [cycle.js](https://github.com/douglascrockford/JSON-js/blob/master/cycle.js) which is supposed to make such things serialisable, but I haven't tried it out just yet. The next thing I tried was creating a blank array, then iterating through `grid.getItems()`, pushing `item.getElement()` to the array every time, and then stringifying the array. Creating the array worked but loading it didn't. I tried tinkering with the custom structure for some time but still couldn't get it to work.

What do I have to do to enable saving to and loading from a file? If I can at the very least get the HTML code for every item, I can use that to initialise the grid's items.

**Comment by niklasramo** [comment]
Have you seen this collection of Muuri demos: https://codepen.io/collection/AWopag/? There's this example: https://codepen.io/niklasramo/pen/YveqNN which should help you get started ;)

**Comment by aorozsa** [comment]
I have seen that demo, and it's not what I'm after. My use case is, I want to be able to save the items in the grid, then, when the grid is of a different size, it would be able to be loaded. I've already tried that sorting example, and it only works for grids of constant sizes; it doesn't work for me. Another important aspect of my use case is that I'm actually creating a desktop application using Electron, so I can't rely on sessions to save/load (unless it actually works with an electron app...)

**Comment by aorozsa** [comment]
Another thing: I need to be able to transfer layouts from one system to another. So... saving to and loading from files.

**Comment by niklasramo** [comment]
Well in that demo you can just store the layout on a variable, it's just a string. But... the problem in your code is (most likely) that you store the grid items on save action, but on load action you remove all the item instances which also _destroys_ the items which makes them unusable (references to DOM nodes are lost). Instead you should store the DOM nodes in the order you wish on save.

**Comment by aorozsa** [comment]
Okay... so how do I save these DOM nodes? Also, this is very early code, so obviously there's refinement that needs to be done.

**Comment by niklasramo** [comment]
`var elements = grid.getItems().map(item => item.getElement())`

**Comment by aorozsa** [comment]
I think I've already tried that without success. I'll try again and post my result.

**Comment by aorozsa** [comment]
Okay, so I have tried it before, but I was doing it wrong since I was trying to stringify it. Once I added that to my code, it had the same results. I printed the type of it to the console, which returned "Object". Can this really be saved to a file and then successfully read?

**Comment by niklasramo** [comment]
Umm no, if you're saving on file you'd probably want to follow the example in the demo I provided. It saves an array of item ids as a string.

**Comment by aorozsa** [comment]
How would that translate into what I need? That example saves ids, and then upon loading, the grid's items that are always there get sorted to match the order in the array. I need to be able to load in items that don't exist. Is this not possible?

**Comment by niklasramo** [comment]
Well, you need to create the DOM nodes for the items first ;) Muuri does not generate DOM nodes automatically (except for the placeholder, but that's customizable).

**Comment by aorozsa** [comment]
I think I'm getting somewhere with this. I'm using [FileSaver.js](https://github.com/eligrey/FileSaver.js/blob/master/src/FileSaver.js) to save to a file, and the prototype code below almost captures the card div elements perfectly. Below is the code, and after that is the output for saving with my grid's initial item (that's even written into the HTML file itself (will probably change that), and another item that was added.

`saveBtn.addEventListener('click', function(event) {`
`  savedItems = grid.getItems().map(item => item.getElement());`
`  var itemsToSave = [];`
`  savedItems.forEach(function(item) {`
`    itemsToSave.push(item.innerHTML);`
`  });`
`  var blob = new Blob([itemsToSave], {type: "text/plain"});`
`  saveAs(blob, "SaveFile.txt");`
`});`

**Comment by aorozsa** [comment]
Alright. I got it working. I even implemented autosaving, which also works with Electron. There are still a couple of minor bugs that need to be ironed out, but the basic idea is there. Thanks for your help @niklasramo :). For people who are also wondering how to do this, I'll briefly walk you through how to do it. The basic idea behind saving are these functions:

`saveBtn.addEventListener('click', function(event) {`
`  var file = new Blob([gridToJSON()], {type: 'application/octet-stream'});`
`  saveAs(file, "filename");`
`});`

`function gridToJSON() {`
`  var items = grid.getItems().map(item => item.getElement());`
`  var itemsToSave = [];`
`  items.forEach(function(item) {`
`    itemsToSave.push(item.firstElementChild.innerHTML);`
`  });`
`  return JSON.stringify(itemsToSave);`
`}`

This simply creates a string of an array which contains part of the items' HTML code (I'll explain a bit more when describing the loading part) and then uses a Blob and [FileSaver.js](https://github.com/eligrey/FileSaver.js/blob/master/src/FileSaver.js) to save to a file. I used `type: 'application/octlet-stream'` for consistency in Electron since saveAs brings up a dialogue for saving, and if you have 'application/json' instead, changing the filename from the default assigned one would save it as an octlet stream. This type is still perfectly readable.


As for loading, these are the functions behind it: 

`loadBtn.addEventListener('change', function(e) {`
`  var file = e.target.files[0];`
`  if (!file) return;`
`  var reader = new FileReader();`
`  reader.onload = function(e) {`
`    var contents = e.target.result;`
`    grid.remove(grid.getItems(), {removeElements: true});`
`    load(contents);`
`  };`
`  reader.readAsText(file);`
`}), false;`

`function load(layout) { // Loads cards that have already been created before`
`  var itemsToLoad = JSON.parse(layout);`
`  itemsToLoad.forEach(function(item) {`
`    addExistingCard(item);`
`  });`
`  window.localStorage.setItem('layout', gridToJSON()); // Saves the grid's items`
`}`

`function addExistingCard(cardDiv) {`
`  var itemElem = document.createElement('div');`
`  var itemTemplate =`
`    '<div class="item">' +`
`    '<div class="item-content">' +`
`    cardDiv +`
`    '</div>' +`
`    '</div>';`
`  itemElem.innerHTML = itemTemplate;`
`  grid.add(itemElem.firstChild);`
`  if (cardDiv.includes("ghost")) {`
`    ghost = grid.getItems(0)[0];`
`  }`
`}`

When I said earlier that there were some bugs that needed to be ironed out, the 'change' event listener is the cause for one of them; it stops the same file from being loaded multiple times in a row. That's because in the markup, loadBtn is actually a label that's linked to a hidden file input, adding other event listeners didn't work, and that bug flew under my radar when I first committed these changes. I'll fix it, I just don't want to right now.
Anyway, the code in the event listener reads the file, sends the output to `load()` to convert it back to an array, and then it iterates through every element, calling `addExistingCard()` to add the card in based off a template.
Before I said that saving creates a string of an array which contains **part** of the items' HTML code, well, the structure my items' HTML follows is this (as mostly laid out in addExistingCard):
`<div class="item"><div class="item-content><div class="card">*card content*  3 -> </div>`.

`item.firstElementChild.innerHTML` copies the "card" div, which is saved to a file, then used in the above loading template to bring back the file. This is purely specific to this card layout, and you'll have to do some digging and testing for how to save your items' HTML code if your card layout is different.
